File: server/lib/store/json_store.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as p;
import 'package:minister_shared/models/account.dart';
import 'package:minister_shared/models/transaction.dart';

late final String dataDir;
late final String serverRoot;

void initStore(String root) {
  serverRoot = root;
  dataDir = p.join(serverRoot, 'data');
  Directory(dataDir).createSync(recursive: true);
}

String get _accountFile => p.join(dataDir, 'linked_account.json');
String get _txFile => p.join(dataDir, 'transactions.json');
String get _cleanFile => p.join(dataDir, 'transactions_clean.json');
String get _overridesFile => p.join(dataDir, 'category_overrides.json');
String get _categoryRulesFile => p.join(dataDir, 'category_rules.json');
String get _balancesFile => p.join(dataDir, 'balances.json');
String get _pinnedTransactionsFile => p.join(dataDir, 'pinned_transactions.json');
String get _deletedDefaultsFile => p.join(dataDir, 'deleted_defaults.json');
String get _defaultCategoryRulesFile => p.join(serverRoot, 'default_category_rules.json');
String get _exampleDefaultCategoryRulesFile => p.join(serverRoot, 'example_default_category_rules.json');

AccountData readAccountData() {
  try {
    final content = File(_accountFile).readAsStringSync();
    return AccountData.fromJson(jsonDecode(content) as Map<String, dynamic>);
  } catch (_) {
    return AccountData();
  }
}

void writeAccountData(AccountData data) {
  Directory(dataDir).createSync(recursive: true);
  File(_accountFile).writeAsStringSync(
      const JsonEncoder.withIndent('  ').convert(data.toJson()));
}

List<StoredTransaction> loadTransactions() {
  try {
    final content = File(_txFile).readAsStringSync();
    return (jsonDecode(content) as List<dynamic>)
        .map((e) => StoredTransaction.fromJson(e as Map<String, dynamic>))
        .toList();
  } catch (_) {
    return [];
  }
}

void saveTransactions(List<StoredTransaction> transactions) {
  Directory(dataDir).createSync(recursive: true);
  File(_txFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ')
        .convert(transactions.map((t) => t.toJson()).toList()),
  );
}

List<CleanTransaction> loadCleanTransactions() {
  try {
    final content = File(_cleanFile).readAsStringSync();
    return (jsonDecode(content) as List<dynamic>)
        .map((e) => CleanTransaction.fromJson(e as Map<String, dynamic>))
        .toList();
  } catch (_) {
    return [];
  }
}

void saveCleanTransactions(List<CleanTransaction> transactions) {
  Directory(dataDir).createSync(recursive: true);
  File(_cleanFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ')
        .convert(transactions.map((t) => t.toJson()).toList()),
  );
}

Map<String, String> loadOverrides() {
  try {
    final content = File(_overridesFile).readAsStringSync();
    return (jsonDecode(content) as Map<String, dynamic>)
        .map((k, v) => MapEntry(k, v as String));
  } catch (_) {
    return {};
  }
}

void saveOverrides(Map<String, String> overrides) {
  Directory(dataDir).createSync(recursive: true);
  File(_overridesFile)
      .writeAsStringSync(const JsonEncoder.withIndent('  ').convert(overrides));
}

Set<String> loadPinnedTransactions() {
  try {
    final content = File(_pinnedTransactionsFile).readAsStringSync();
    return (jsonDecode(content) as List<dynamic>).cast<String>().toSet();
  } catch (_) {
    return {};
  }
}

void savePinnedTransactions(Set<String> pinned) {
  Directory(dataDir).createSync(recursive: true);
  File(_pinnedTransactionsFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert(pinned.toList()),
  );
}

String accountLabel(LinkedAccount acct) {
  return [
    acct.institution,
    acct.displayName,
    acct.last4 != null ? '****${acct.last4}' : null
  ].where((s) => s != null && s.isNotEmpty).join(' ');
}

class CategoryRule {
  final String id;
  final String category;
  final String pattern;
  final bool caseSensitive;

  CategoryRule({
    required this.id,
    required this.category,
    required this.pattern,
    this.caseSensitive = false,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'category': category,
        'pattern': pattern,
        'caseSensitive': caseSensitive,
      };

  factory CategoryRule.fromJson(Map<String, dynamic> json) => CategoryRule(
        id: json['id'] as String? ?? DateTime.now().millisecondsSinceEpoch.toString(),
        category: json['category'] as String,
        pattern: json['pattern'] as String,
        caseSensitive: json['caseSensitive'] as bool? ?? false,
      );

  RegExp toRegExp() => RegExp(pattern, caseSensitive: caseSensitive);
}

List<CategoryRule> loadCategoryRules() {
  try {
    final content = File(_categoryRulesFile).readAsStringSync();
    return (jsonDecode(content) as List<dynamic>)
        .map((e) => CategoryRule.fromJson(e as Map<String, dynamic>))
        .toList();
  } catch (_) {
    return [];
  }
}

void saveCategoryRules(List<CategoryRule> rules) {
  Directory(dataDir).createSync(recursive: true);
  File(_categoryRulesFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ')
        .convert(rules.map((r) => r.toJson()).toList()),
  );
}

Map<String, dynamic> loadBalances() {
  try {
    final content = File(_balancesFile).readAsStringSync();
    return jsonDecode(content) as Map<String, dynamic>;
  } catch (_) {
    return {};
  }
}

void saveBalances(Map<String, dynamic> balances) {
  Directory(dataDir).createSync(recursive: true);
  File(_balancesFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert(balances),
  );
}

Set<String> loadDeletedDefaults() {
  try {
    final content = File(_deletedDefaultsFile).readAsStringSync();
    return (jsonDecode(content) as List<dynamic>).cast<String>().toSet();
  } catch (_) {
    return {};
  }
}

void saveDeletedDefaults(Set<String> deletedDefaults) {
  Directory(dataDir).createSync(recursive: true);
  File(_deletedDefaultsFile).writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert(deletedDefaults.toList()),
  );
}

List<CategoryRule> loadDefaultCategoryRules() {
  // Initialize default_category_rules.json from example if it doesn't exist
  final defaultFile = File(_defaultCategoryRulesFile);
  if (!defaultFile.existsSync()) {
    final exampleFile = File(_exampleDefaultCategoryRulesFile);
    if (exampleFile.existsSync()) {
      defaultFile.writeAsStringSync(exampleFile.readAsStringSync());
    } else {
      // If example doesn't exist, return empty list
      return [];
    }
  }

  try {
    final content = defaultFile.readAsStringSync();
    return (jsonDecode(content) as List<dynamic>)
        .map((e) => CategoryRule.fromJson(e as Map<String, dynamic>))
        .toList();
  } catch (_) {
    return [];
  }
}

File: server/src/store/jsonStore.ts
import fs from 'fs';
import path from 'path';
import type { AccountData, CategoryRule, CleanTransaction, LinkedAccount, StoredTransaction } from '../types.js';

let dataDir = '';
let serverRoot = '';

export function initStore(root: string): void {
  serverRoot = root;
  dataDir = path.join(root, 'data');
  fs.mkdirSync(dataDir, { recursive: true });
}

function accountFile(): string { return path.join(dataDir, 'linked_account.json'); }
function txFile(): string { return path.join(dataDir, 'transactions.json'); }
function cleanFile(): string { return path.join(dataDir, 'transactions_clean.json'); }
function overridesFile(): string { return path.join(dataDir, 'category_overrides.json'); }
function categoryRulesFile(): string { return path.join(dataDir, 'category_rules.json'); }
function balancesFile(): string { return path.join(dataDir, 'balances.json'); }
function pinnedTransactionsFile(): string { return path.join(dataDir, 'pinned_transactions.json'); }
function deletedDefaultsFile(): string { return path.join(dataDir, 'deleted_defaults.json'); }
function defaultCategoryRulesFile(): string { return path.join(serverRoot, 'default_category_rules.json'); }
function exampleDefaultCategoryRulesFile(): string { return path.join(serverRoot, 'example_default_category_rules.json'); }

function readJson<T>(file: string, fallback: T): T {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf-8')) as T;
  } catch {
    return fallback;
  }
}

function writeJson(file: string, data: unknown): void {
  fs.mkdirSync(dataDir, { recursive: true });
  fs.writeFileSync(file, JSON.stringify(data, null, 2));
}

export function accountLabel(acct: LinkedAccount): string {
  return [acct.institution, acct.display_name, acct.last4 ? `****${acct.last4}` : undefined]
    .filter((s): s is string => typeof s === 'string' && s.length > 0)
    .join(' ');
}

// ── Account data ──────────────────────────────────────────────────────────────

export function readAccountData(): AccountData {
  return readJson<AccountData>(accountFile(), { accounts: [] });
}

export function writeAccountData(data: AccountData): void {
  writeJson(accountFile(), data);
}

// ── Transactions ──────────────────────────────────────────────────────────────

export function loadTransactions(): StoredTransaction[] {
  return readJson<StoredTransaction[]>(txFile(), []);
}

export function saveTransactions(txs: StoredTransaction[]): void {
  writeJson(txFile(), txs);
}

export function loadCleanTransactions(): CleanTransaction[] {
  return readJson<CleanTransaction[]>(cleanFile(), []);
}

export function saveCleanTransactions(txs: CleanTransaction[]): void {
  writeJson(cleanFile(), txs);
}

// ── Overrides ─────────────────────────────────────────────────────────────────

export function loadOverrides(): Record<string, string> {
  return readJson<Record<string, string>>(overridesFile(), {});
}

export function saveOverrides(overrides: Record<string, string>): void {
  writeJson(overridesFile(), overrides);
}

// ── Pinned transactions ───────────────────────────────────────────────────────

export function loadPinnedTransactions(): Set<string> {
  const arr = readJson<string[]>(pinnedTransactionsFile(), []);
  return new Set(arr);
}

export function savePinnedTransactions(pinned: Set<string>): void {
  writeJson(pinnedTransactionsFile(), [...pinned]);
}

// ── Category rules ────────────────────────────────────────────────────────────

export function loadCategoryRules(): CategoryRule[] {
  return readJson<CategoryRule[]>(categoryRulesFile(), []);
}

export function saveCategoryRules(rules: CategoryRule[]): void {
  writeJson(categoryRulesFile(), rules);
}

// ── Balances ──────────────────────────────────────────────────────────────────

export function loadBalances(): Record<string, unknown> {
  return readJson<Record<string, unknown>>(balancesFile(), {});
}

export function saveBalances(balances: Record<string, unknown>): void {
  writeJson(balancesFile(), balances);
}

// ── Deleted defaults ──────────────────────────────────────────────────────────

export function loadDeletedDefaults(): Set<string> {
  const arr = readJson<string[]>(deletedDefaultsFile(), []);
  return new Set(arr);
}

export function saveDeletedDefaults(deleted: Set<string>): void {
  writeJson(deletedDefaultsFile(), [...deleted]);
}

// ── Default category rules ────────────────────────────────────────────────────

export function loadDefaultCategoryRules(): CategoryRule[] {
  const defaultFile = defaultCategoryRulesFile();
  if (!fs.existsSync(defaultFile)) {
    const exampleFile = exampleDefaultCategoryRulesFile();
    if (fs.existsSync(exampleFile)) {
      fs.writeFileSync(defaultFile, fs.readFileSync(exampleFile, 'utf-8'));
    } else {
      return [];
    }
  }
  return readJson<CategoryRule[]>(defaultFile, []);
}
